task_manager/task_manager.hh

```diff
@@ -0,0 +1,308 @@
+    using foreign_task_vector = std::vector<foreign_task_ptr>;
+    using module_ptr = seastar::lw_shared_ptr<module>;
+    using modules = std::unordered_map<std::string, module_ptr>;
+private:
+    task_map _all_tasks;
+    modules _modules;
+    config _cfg;
+    abort_source& _as;
+    serialized_action _update_task_ttl_action;
+    utils::observer<uint32_t> _task_ttl_observer;
+    uint32_t _task_ttl;
+public:
+    class task_not_found : public std::exception {
+        sstring _cause;
+    public:
+        explicit task_not_found(task_id tid)
+            : _cause(format("task with id {} not found", tid))
+        { }
+
+        virtual const char* what() const noexcept override { return _cause.c_str(); }
+    };
+
+    enum class task_state {
+        created,
+        running,
+        done,
+        failed
+    };
+
+    class task {
+    public:
+        struct progress {
+            double completed = 0.0;         // Number of units completed so far.
+            double total = 0.0;             // Total number of units to complete the task.
+        };
+
+        struct status {
+            task_id id;
+            std::string type;
+            task_state state = task_state::created;
+            db_clock::time_point start_time;
+            db_clock::time_point end_time;
+            std::string error;
+            int sequence_number = 0;        // A running sequence number of the task.
+            unsigned shard = 0;
+            std::string keyspace;
+            std::string table;
+            std::string entity;             // Additional entity specific for the given type of task.
+            std::string progress_units;     // A description of the units progress.
+        };
+
+        class impl {
+        protected:
+            status _status;
+            progress _progress;             // Reliable only for tasks with no descendants.
+            foreign_task_ptr _parent;
+            foreign_task_vector _children;
+            seastar::shared_promise<> _done;
+            module_ptr _module;
+        public:
+            impl(module_ptr module, task_id id, unsigned sequence_number, std::string keyspace, std::string table, std::string type, std::string entity, foreign_task_ptr&& parent) 
+                : _status({
+                    .id = id,
+                    .type = std::move(type),
+                    .state = task_state::created,
+                    .sequence_number = sequence_number,
+                    .shard = this_shard_id(),
+                    .keyspace = std::move(keyspace),
+                    .table = std::move(table),
+                    .entity = std::move(entity)
+                    })
+                , _parent(std::move(parent))
+                , _module(module)
+            {}
+
+            virtual future<progress> get_progress() const = 0;
+            virtual is_abortable is_abortable() const noexcept = 0;
+            virtual future<> abort() = 0;
```

Comment:
Can have sensible default implementations of the above virtual methods to reduce the boilerplate

Reply:
Why not implement abort as a noop, returning `make_ready_future<>()`?
If a task impl is not abortable, why forcing it to implement `abort()`?

Also, let's declare it noexcept.
Dealing with exception on the abort / shutdown path is a nightmare.
There's nothing we can do with them at this point.